src/server/oasisapi/analyses/models.py

```diff
@@ -205,18 +209,30 @@ def generate_input_signature(self):
         info_file = self.portfolio.reinsurance_info_file.file.name if self.portfolio.reinsurance_info_file else None
         scope_file = self.portfolio.reinsurance_source_file.file.name if self.portfolio.reinsurance_source_file else None
         settings_file = self.settings_file.file.name if self.settings_file else None
+        complex_data_files = self.create_complex_model_data_file_dicts()
 
         return signature(
             'generate_input',
-            args=(loc_file, acc_file, info_file, scope_file, settings_file),
+            args=(loc_file, acc_file, info_file, scope_file, settings_file, complex_data_files),
             queue=self.model.queue_name
         )
 
+    def create_complex_model_data_file_dicts(self):
+        """Creates a list of tuples containing metadata for the complex model data files.
+
+        Returns:
+            list of dict: Dicts containing (stored filename, original filename) as the keys.
+
+        """
+        complex_data_files = [{"stored_filename": cmdf.data_file.file.name,
+                               "original_filename": cmdf.file_name}
+                              for cmdf in self.complex_model_data_files.all()]
+        return complex_data_files
+
     def copy(self):
         new_instance = self
         new_instance.pk = None
         new_instance.name = '{} - Copy'.format(new_instance.name)
-        new_instance.creator = None
```

Comment:
The reasoning for this one is a bit roundabout. It comes up because of the addition of the ComplexModelDataFile as a Many-to-Many field on the Analyses model. The sequence of problems is:

1. When copying an Analysis via the API, it is possible to assign new ComplexModelDataFiles. However, Django raised an error, as adding a Many-to-Many entry requires both objects to have database IDs, but the newly copied Analysis did not have one, as it had not been saved.
2. The Analysis cannot be saved without a Creator, as it is a mandatory field. This was previously handled by the Serializer, which when validating the Analysis, adds a Creator as necessary, and the Analysis was only saved after.
3. However, the Serializer also handles the step of applying the new CMDF associations - so the Analysis couldn't be saved before calling it due to the invalid missing data, and couldn't be saved after as Django tried save the Many-to-Many relations referring to an as-yet unsaved Analysis!

My workaround is to leave the creator populated here, and then in analyses.viewsets.copy the Analysis is saved to generate an ID, after which the creator is wiped and re-populated.